gcsfs/core.py

```diff
@@ -1086,54 +1086,70 @@ async def _isdir(self, path):
     async def _find(self, path, withdirs=False, detail=False, prefix="", **kwargs):
         path = self._strip_protocol(path)
         bucket, key = self.split_path(path)
-        out, _ = await self._do_list_objects(
-            path,
-            delimiter=None,
-            prefix=prefix,
-        )
-        if not prefix and not out and key:
-            try:
-                out = [
-                    await self._get_object(
-                        path,
-                    )
-                ]
-            except FileNotFoundError:
-                out = []
-        dirs = []
-        sdirs = set()
+
+        if prefix:
+            _path = "" if not key else key.rstrip("/") + "/"
+            _prefix = f"{_path}{prefix}"
+        else:
+            _prefix = key
+
+        objects, _ = await self._do_list_objects(bucket, delimiter="", prefix=_prefix)
+
+        dirs = {}
         cache_entries = {}
-        for o in out:
-            par = o["name"]
-            while par:
-                par = self._parent(par)
-                if par not in sdirs:
-                    if len(par) < len(path):
-                        break
-                    sdirs.add(par)
-                    dirs.append(
-                        {
-                            "Key": self.split_path(par)[1],
-                            "Size": 0,
-                            "name": par,
-                            "StorageClass": "DIRECTORY",
-                            "type": "directory",
-                            "size": 0,
-                        }
-                    )
-                # Don't cache "folder-like" objects (ex: "Create Folder" in GCS console) to prevent
+
+        for obj in objects:
+            obj_parent = self._parent(obj["name"])
+
+            parent = obj["name"]
+            while parent:
+                parent = self._parent(parent)
+                if len(parent) < len(path):
+                    # don't go above the requested level
+                    break
+
+                dir_key = self.split_path(parent)[1]
+                if not dir_key:
+                    break
+
+                if parent in dirs:
+                    continue
+
+                dir_info = {
+                    "Key": dir_key,
+                    "Size": 0,
+                    "name": parent,
+                    "StorageClass": "DIRECTORY",
+                    "type": "directory",
+                    "size": 0,
+                }
+
+                dirs[parent] = dir_info
+                pparent = self._parent(parent)
+                if len(pparent) < len(path):
+                    continue
+                cache_entries.setdefault(pparent, []).append(dir_info)
```

Comment:
The number of parents is confusing here. `obj_parent`, `parent`, `pparent`. Also looks like the 
```
if len(pparent) < len(path):
                    continue
```

is equivalent to breaking, right? We've discussed this before, but maybe there is a way to extract a helper out of it, to break up the context and make it more digestable, since we are already getting into renaming stuff.

Reply:
Got rid of parents and tried to simplify even more

Reply:
Looks much better now, thank you! 🔥